File: tokenization/bench.py
# /// script
# requires-python = ">=3.13"
# dependencies = [
#   "stringzilla>=4.5.0",
#   "regex",
#   "PyICU",
# ]
# ///
"""
Python UTF-8 tokenization and iteration benchmarks.

Benchmarks segmentation and codepoint operations:
- TR29 word segmentation (lazy iterators), per document line
- Unicode whitespace and newline splitting, per document line
- UTF-8 codepoint counting, over the whole document

The splitter benches process one document line per call, cycling through the lines (default
tokenization mode is 'lines'); the codepoint counter scans the whole document buffer.

Environment variables:
- STRINGWARS_DATASET: Path to input dataset file
- STRINGWARS_TOKENS: Tokenization mode ('lines', 'words', 'file'); defaults to 'lines'

Examples:
  uv run tokenization/bench.py --dataset README.md
  STRINGWARS_DATASET=data.txt uv run tokenization/bench.py

Timing via time.monotonic_ns; throughput in decimal GB/s. Filter with -k/--filter.
"""

import argparse
import itertools
import re
import sys
from collections.abc import Callable
from importlib.metadata import version as pkg_version

import icu
import regex
import stringzilla as sz

from utils import (
    add_common_args,
    load_dataset,
    now_nanoseconds,
    paced_items,
    report_stats,
    resolve_tokens,
    should_run,
    tokenize_dataset,
)


def log_system_info():
    """Log Python version and library versions."""
    print(f"- Python: {sys.version.split()[0]}, {sys.platform}")
    print(f"- StringZilla: {sz.__version__} with {sz.__capabilities_str__}")
    print(f"- regex: {pkg_version('regex')}")
    print(f"- PyICU: {pkg_version('PyICU')} (ICU {icu.ICU_VERSION})")
    print()


def bench_tokenize(
    name: str,
    text: str | bytes,
    count_function: Callable[[str | bytes], int],
    time_limit_seconds: float = 10.0,
):
    """Benchmark a whole-text tokenizer/scanner by counting what it yields.

    `count_function` consumes the entire `text` once per call and returns an integer
    count (token count, codepoint count, ...). The StringZilla and regex/ICU paths
    count lazily via `sum(1 for _ in ...)` so no token list is materialized; the
    stdlib `str.split()/splitlines()` paths do allocate a list, as that is the only
    idiom they offer. Throughput is reported as input bytes per second.
    """
    text_byte_length = len(text.encode("utf-8")) if isinstance(text, str) else len(text)

    start_time = now_nanoseconds()
    deadline_nanoseconds = start_time + int(time_limit_seconds * 1e9)

    passes = 0
    token_total = 0
    # Each yielded sentinel triggers one full-text pass; a pass is expensive, so we
    # check the deadline every iteration (step=1) rather than overshoot a large file.
    for _ in paced_items(itertools.count(), deadline_nanoseconds, step=1):
        token_total += count_function(text)
        passes += 1

    seconds = (now_nanoseconds() - start_time) / 1e9
    tokens_per_pass = token_total // passes if passes else 0

    print(f"{name}: {tokens_per_pass:,} tokens per pass over {passes:,} passes", file=sys.stderr)
    report_stats(name, "bytes", seconds, passes, text_byte_length * passes)


def bench_split_lines(
    name: str,
    lines: list[str],
    count_function: Callable[[str], int],
    time_limit_seconds: float = 10.0,
):
    """Benchmark a splitter by processing one document line per call, cycling the lines.

    Unlike the whole-text `bench_tokenize`, each call splits a single line, so the working set
    is one line rather than the entire file. Splitting is compute-bound, so the per-byte rate
    still mirrors a whole-file pass; only the working set changes. Throughput is reported as the
    sum of the byte lengths of every line processed.
    """
    line_byte_lengths = [len(line.encode("utf-8")) for line in lines]

    start_time = now_nanoseconds()
    deadline_nanoseconds = start_time + int(time_limit_seconds * 1e9)

    lines_processed = 0
    token_total = 0
    bytes_total = 0
    for line, line_byte_length in paced_items(
        itertools.cycle(zip(lines, line_byte_lengths, strict=True)), deadline_nanoseconds
    ):
        token_total += count_function(line)
        bytes_total += line_byte_length
        lines_processed += 1

    seconds = (now_nanoseconds() - start_time) / 1e9
    tokens_per_line = token_total / lines_processed if lines_processed else 0

    print(
        f"{name}: {tokens_per_line:,.1f} tokens per line over {lines_processed:,} lines",
        file=sys.stderr,
    )
    report_stats(name, "bytes", seconds, lines_processed, bytes_total)


def count_words_stringzilla(text: str) -> int:
    """Count TR29 words lazily via StringZilla's utf8_word_iter (no list built)."""
    return sum(1 for _ in sz.utf8_word_iter(text, skip_empty=True))


def count_words_regex(text: str) -> int:
    """Count word-like runs via the `regex` module's \\w+ finditer (lazy)."""
    return sum(1 for _ in regex.finditer(r"\w+", text))


def count_words_split(text: str) -> int:
    """Count whitespace-delimited words via stdlib str.split() (allocates a list)."""
    return len(text.split())


def make_count_words_icu() -> Callable[[str], int]:
    """Build an ICU word BreakIterator counter, reusing one iterator instance.

    Iterating the break iterator yields every boundary segment (words, punctuation,
    and whitespace), mirroring the Rust ICU WordSegmenter baseline which also counts
    boundary segments rather than only word-like ones.
    """
    break_iterator = icu.BreakIterator.createWordInstance(icu.Locale.getRoot())

    def count(text: str) -> int:
        break_iterator.setText(text)
        segments = 0
        for _ in break_iterator:
            segments += 1
        return segments

    return count


def count_whitespace_stringzilla(text: str) -> int:
    """Count whitespace-delimited tokens lazily via StringZilla's utf8_split_iter."""
    return sum(1 for _ in sz.utf8_split_iter(text, skip_empty=True))


def count_whitespace_split(text: str) -> int:
    """Count whitespace-delimited tokens via stdlib str.split() (allocates a list)."""
    return len(text.split())


def count_newlines_stringzilla(text: str) -> int:
    """Count lines lazily via StringZilla's utf8_splitlines_iter (no list built)."""
    return sum(1 for _ in sz.utf8_splitlines_iter(text))


def count_newlines_splitlines(text: str) -> int:
    """Count lines via stdlib str.splitlines() (allocates a list)."""
    return len(text.splitlines())


def count_codepoints_stringzilla(data: bytes) -> int:
    """Count UTF-8 codepoints in raw bytes via StringZilla's utf8_count (SIMD scan)."""
    return sz.utf8_count(data)


def count_codepoints_decode(data: bytes) -> int:
    """Count UTF-8 codepoints by decoding the bytes to a str, then len() (allocates)."""
    return len(data.decode("utf-8"))


_main_epilog = """
Examples:

  # Benchmark all tokenization and iteration operations
  %(prog)s --dataset README.md --tokens file

  # Test only word segmentation
  %(prog)s --dataset data.txt --tokens file -k "words"
"""


def main():
    """Main entry point with argument parsing."""
    parser = argparse.ArgumentParser(
        description="Benchmark UTF-8 tokenization and iteration",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=_main_epilog,
    )

    add_common_args(parser)

    args = parser.parse_args()

    # Compile filter pattern
    filter_pattern: re.Pattern | None = None
    if args.filter:
        try:
            filter_pattern = re.compile(args.filter)
        except re.error as e:
            parser.error(f"Invalid regex for --filter: {e}")

    # Load the dataset and tokenize it into document lines (the per-line splitter work items).
    tokens_mode = resolve_tokens(args.tokens, "lines")
    pythonic_str = load_dataset(args.dataset, as_bytes=False, size_limit=args.dataset_limit)
    lines = tokenize_dataset(pythonic_str, tokens_mode)

    if not lines:
        print("No tokens found in dataset")
        return 1

    total_tokens = len(lines)
    mean_token_length = sum(len(t) for t in lines) / total_tokens
    total_bytes = len(pythonic_str)

    print(f"Dataset: {total_tokens:,} tokens, {total_bytes:,} bytes, {mean_token_length:.1f} avg token length")
    log_system_info()

    # UTF-8 word segmentation (TR29) per document line, cycling the lines.
    print("Word Segmentation (TR29)")
    if should_run("tokenize-words/stringzilla.utf8_word_iter()", filter_pattern):
        bench_split_lines("stringzilla.utf8_word_iter()", lines, count_words_stringzilla, args.time_limit)
    if should_run("tokenize-words/regex.finditer(\\w+)", filter_pattern):
        bench_split_lines("regex.finditer(\\w+)", lines, count_words_regex, args.time_limit)
    if should_run("tokenize-words/icu.BreakIterator()", filter_pattern):
        bench_split_lines("icu.BreakIterator()", lines, make_count_words_icu(), args.time_limit)
    if should_run("tokenize-words/std.str.split()", filter_pattern):
        bench_split_lines("std.str.split()", lines, count_words_split, args.time_limit)

    # UTF-8 whitespace splitting per document line, cycling the lines.
    print("\nWhitespace Splitting")
    if should_run("tokenize-whitespace/stringzilla.utf8_split_iter()", filter_pattern):
        bench_split_lines("stringzilla.utf8_split_iter()", lines, count_whitespace_stringzilla, args.time_limit)
    if should_run("tokenize-whitespace/std.str.split()", filter_pattern):
        bench_split_lines("std.str.split()", lines, count_whitespace_split, args.time_limit)

    # UTF-8 newline splitting per document line, cycling the lines.
    print("\nNewline Splitting")
    if should_run("tokenize-newlines/stringzilla.utf8_splitlines_iter()", filter_pattern):
        bench_split_lines("stringzilla.utf8_splitlines_iter()", lines, count_newlines_stringzilla, args.time_limit)
    if should_run("tokenize-newlines/std.str.splitlines()", filter_pattern):
        bench_split_lines("std.str.splitlines()", lines, count_newlines_splitlines, args.time_limit)

    # UTF-8 codepoint counting over the raw bytes (fair O(n)-from-bytes comparison;
    # `len(str)` is O(1) in CPython, so we decode-and-count as the stdlib baseline).
    print("\nCodepoint Counting")
    document_bytes = pythonic_str.encode("utf-8")
    if should_run("utf8-count/stringzilla.utf8_count()", filter_pattern):
        bench_tokenize("stringzilla.utf8_count()", document_bytes, count_codepoints_stringzilla, args.time_limit)
    if should_run("utf8-count/std.decode().len()", filter_pattern):
        bench_tokenize("std.decode().len()", document_bytes, count_codepoints_decode, args.time_limit)

    return 0


if __name__ == "__main__":
    exit(main())

File: tokenization/bench.rs
#![doc = r#"
# StringWars: UTF-8 Tokenization & Iteration Benchmarks

This file benchmarks UTF-8 segmentation and codepoint iteration:
- Unicode whitespace and newline splitting
- Unicode TR29 (UAX#29) word segmentation
- UTF-8 character counting and UTF-8 to UTF-32 decoding
- Locating the byte offset of the Nth UTF-8 codepoint

## Benchmark Groups

- `tokenize-whitespace`: Unicode whitespace splitting
- `tokenize-newlines`: Unicode newline splitting
- `tokenize-words-tr29`: Unicode TR29 word boundary segmentation
- `utf8-length`: UTF-8 character counting
- `utf8-iterate`: UTF-8 to UTF-32 decoding
- `find-nth-utf8`: byte offset of the Nth UTF-8 codepoint

## Usage Examples

```sh
RUSTFLAGS="-C target-cpu=native" \
    STRINGWARS_DATASET=README.md \
    STRINGWARS_TOKENS=lines \
    cargo bench --features bench_tokenization --bench bench_tokenization
```
"#]
use std::hint::black_box;

use stringtape::BytesCowsAuto;

use icu::properties::props::WhiteSpace;
use icu::properties::CodePointSetData;
use icu::segmenter::WordSegmenter;
use stringzilla::sz;
use stringzilla::sz::StringZillableUnary;
use unicode_segmentation::UnicodeSegmentation;

#[path = "../utils.rs"]
mod utils;
use utils::{
    install_panic_hook, load_dataset, log_stringzilla_metadata, measure_throughput, should_run,
    BenchBudget, ReportAs, ResultExt, WorkUnits,
};

/// File-local helper: cycles through `needles` byte slices, passes each to `count`, and reports
/// throughput as `WorkUnits::bytes(line.len())` — the bytes of that one line per call.
///
/// Only use this for benchmarks whose body is exactly "pick a line, count something,
/// WorkUnits::bytes(line.len())". Blocks that operate on `&str` slices, use `unsafe`, or scan the
/// whole haystack in one shot are left inline.
fn measure_line_tokenizer<Count: FnMut(&[u8]) -> usize>(
    name: &str,
    budget: &BenchBudget,
    needles: &BytesCowsAuto,
    mut count: Count,
) {
    if !should_run(name) {
        return;
    }
    let mut lines = needles.iter().cycle();
    measure_throughput(name, ReportAs::Bytes, budget, || {
        let line = black_box(lines.next().unwrap());
        black_box(count(line));
        WorkUnits::bytes(line.len() as u64)
    });
}

/// Benchmarks Unicode whitespace splitting using ICU, stdlib, and StringZilla.
///
/// Each call splits a single document line, cycling through the line tokens. Throughput is
/// reported as the bytes of that one line, so the per-byte rate still reflects the splitter's
/// compute cost while the working set stays a single line rather than the whole file.
fn bench_tokenize_whitespace(budget: &BenchBudget, _haystack: &[u8], needles: &BytesCowsAuto) {
    // Pre-decode every line to `&str` once, outside the timed closures. The byte-based StringZilla
    // variant operates on the raw line bytes; the `str`-based baselines reuse these validated lines.
    let lines_str: Vec<&str> = needles
        .iter()
        .map(|line| std::str::from_utf8(line).unwrap_or(""))
        .collect();

    // Benchmark for StringZilla whitespace splits.
    measure_line_tokenizer(
        "tokenize-whitespace/stringzilla::utf8_whitespace_splits",
        budget,
        needles,
        |line| {
            let count: usize = line.sz_utf8_whitespace_splits().count();
            black_box(count);
            count
        },
    );

    // Benchmark for Rust stdlib char::is_whitespace.
    {
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-whitespace/std::split<is_whitespace>",
            ReportAs::Bytes,
            budget,
            || {
                let line = lines.next().unwrap();
                let count: usize = black_box(*line)
                    .split(char::is_whitespace)
                    .filter(|segment| !segment.is_empty())
                    .count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }

    // Benchmark for ICU4X WhiteSpace property.
    {
        let white_space = CodePointSetData::new::<WhiteSpace>();
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-whitespace/icu::WhiteSpace.split",
            ReportAs::Bytes,
            budget,
            || {
                let line = lines.next().unwrap();
                let count: usize = black_box(*line)
                    .split(|character: char| white_space.contains(character))
                    .filter(|segment: &&str| !segment.is_empty())
                    .count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }
}

/// Benchmarks Unicode newline splitting using custom predicates and StringZilla.
///
/// Each call splits a single document line, cycling through the line tokens; throughput is the
/// bytes of that one line. (Per-line newline splitting is degenerate when lines were split on `\n`,
/// but the kernels still exercise the full Unicode newline set across the seven characters.)
fn bench_tokenize_newlines(budget: &BenchBudget, _haystack: &[u8], needles: &BytesCowsAuto) {
    // Custom newline predicate matching StringZilla's 7 newline characters.
    fn is_unicode_newline(character: char) -> bool {
        matches!(
            character,
            '\n' | '\r' | '\x0B' | '\x0C' | '\u{0085}' | '\u{2028}' | '\u{2029}'
        )
    }

    // Pre-decode every line to `&str` once (only the custom `str` baseline needs it); the
    // StringZilla variant splits the raw line bytes directly.
    let lines_str: Vec<&str> = needles
        .iter()
        .map(|line| std::str::from_utf8(line).unwrap_or(""))
        .collect();

    // Benchmark for StringZilla newline splits.
    measure_line_tokenizer(
        "tokenize-newlines/stringzilla::utf8_newline_splits",
        budget,
        needles,
        |line| {
            let count: usize = line.sz_utf8_newline_splits().count();
            black_box(count);
            count
        },
    );

    // Benchmark for custom newline predicate.
    {
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-newlines/custom::split<is_unicode_newline>",
            ReportAs::Bytes,
            budget,
            || {
                let line = lines.next().unwrap();
                let count: usize = black_box(*line)
                    .split(is_unicode_newline)
                    .filter(|segment| !segment.is_empty())
                    .count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }
}

/// Benchmarks Unicode TR29 (UAX#29) word segmentation.
///
/// TR29 defines linguistically-aware word boundaries that handle complex cases like
/// contractions ("can't"), numeric sequences ("3.14"), and scripts without spaces.
/// - `unicode-segmentation::unicode_words()`: Filters to word-like segments only
/// - `unicode-segmentation::split_word_bounds()`: All boundary segments including punctuation
/// - `icu::segmenter::WordSegmenter`: ICU4X implementation with LSTM/dictionary models
fn bench_tokenize_words_tr29(budget: &BenchBudget, _haystack: &[u8], needles: &BytesCowsAuto) {
    // Pre-decode every line to `&str` once. StringZilla segments the raw line bytes directly; the
    // `unicode-segmentation`, ICU, and stdlib baselines reuse these validated lines per call.
    let lines_str: Vec<&str> = needles
        .iter()
        .map(|line| std::str::from_utf8(line).unwrap_or(""))
        .collect();

    // Benchmark for StringZilla's single-pass TR29 word iterator. `.count()` consumes the
    // iterator without materializing the segments, so no allocation taints the measurement.
    measure_line_tokenizer(
        "tokenize-words-tr29/stringzilla::utf8_word_splits",
        budget,
        needles,
        |line| {
            let count: usize = line.sz_utf8_word_splits().count();
            black_box(count);
            count
        },
    );

    // Benchmark for unicode-segmentation: unicode_words() - only word-like segments
    {
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-words-tr29/unicode-segmentation::unicode_words",
            ReportAs::Bytes,
            budget,
            || {
                let line = black_box(*lines.next().unwrap());
                let count: usize = line.unicode_words().count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }

    // Benchmark for unicode-segmentation: split_word_bounds() - all segments
    {
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-words-tr29/unicode-segmentation::split_word_bounds",
            ReportAs::Bytes,
            budget,
            || {
                let line = black_box(*lines.next().unwrap());
                let count: usize = line.split_word_bounds().count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }

    // Benchmark for ICU4X WordSegmenter with dictionary model
    {
        let segmenter = WordSegmenter::new_dictionary(Default::default());
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-words-tr29/icu::WordSegmenter::new_dictionary.segment_str",
            ReportAs::Bytes,
            budget,
            || {
                let line = black_box(*lines.next().unwrap());
                // WordSegmenter returns boundary indices; count segments = boundaries - 1
                let boundaries: usize = segmenter.segment_str(line).count();
                black_box(boundaries);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }

    // Benchmark for stdlib split_whitespace as baseline comparison
    {
        let mut lines = lines_str.iter().cycle();
        measure_throughput(
            "tokenize-words-tr29/std::split_whitespace",
            ReportAs::Bytes,
            budget,
            || {
                let line = black_box(*lines.next().unwrap());
                let count: usize = line.split_whitespace().count();
                black_box(count);
                WorkUnits::bytes(line.len() as u64)
            },
        );
    }
}

/// Benchmarks UTF-8 character counting using StringZilla, simdutf, and stdlib.
fn bench_utf8_length(budget: &BenchBudget, haystack: &[u8], _needles: &BytesCowsAuto) {
    let haystack_length = haystack.len() as u64;

    // Validate UTF-8 once, outside the timed closures (only the stdlib baseline needs it; the
    // StringZilla and simdutf counters operate directly on bytes).
    let haystack_str = std::str::from_utf8(haystack).ok();

    // Benchmark for StringZilla UTF-8 character counting via the lazy view.
    measure_throughput(
        "utf8-length/stringzilla::utf8_chars.len",
        ReportAs::Bytes,
        budget,
        || {
            let haystack_bytes = black_box(haystack);
            let count: usize = haystack_bytes.sz_utf8_chars().len();
            black_box(count);
            WorkUnits::bytes(haystack_length)
        },
    );

    // Benchmark for StringZilla's dedicated `count_utf8()` free function (direct SIMD scan,
    // without constructing a view object).
    measure_throughput(
        "utf8-length/stringzilla::count_utf8",
        ReportAs::Bytes,
        budget,
        || {
            let haystack_bytes = black_box(haystack);
            let count: usize = sz::count_utf8(haystack_bytes);
            black_box(count);
            WorkUnits::bytes(haystack_length)
        },
    );

    // Benchmark for simdutf UTF-8 character counting.
    measure_throughput(
        "utf8-length/simdutf::count_utf8",
        ReportAs::Bytes,
        budget,
        || {
            let haystack_bytes = black_box(haystack);
            let count: usize = simdutf::count_utf8(haystack_bytes);
            black_box(count);
            WorkUnits::bytes(haystack_length)
        },
    );

    // Benchmark for stdlib UTF-8 character counting.
    {
        let text = haystack_str.expect("UTF-8 text required for the stdlib codepoint counter");
        measure_throughput(
            "utf8-length/std::chars.count",
            ReportAs::Bytes,
            budget,
            || {
                let count: usize = black_box(text).chars().count();
                black_box(count);
                WorkUnits::bytes(haystack_length)
            },
        );
    }
}

/// Benchmarks UTF-8 to UTF-32 decoding using StringZilla, simdutf, and stdlib.
fn bench_utf8_iterate(budget: &BenchBudget, haystack: &[u8], _needles: &BytesCowsAuto) {
    let haystack_length = haystack.len() as u64;

    // Benchmark for StringZilla UTF-8 character iteration.
    measure_throughput(
        "utf8-iterate/stringzilla::utf8_chars.iter",
        ReportAs::Bytes,
        budget,
        || {
            let haystack_bytes = black_box(haystack);
            let mut sum: u32 = 0;
            for character in haystack_bytes.sz_utf8_chars().iter() {
                sum = sum.wrapping_add(character as u32);
            }
            black_box(sum);
            WorkUnits::bytes(haystack_length)
        },
    );

    // Benchmark for simdutf UTF-8 to UTF-32 conversion.
    {
        // Pre-allocate buffer for UTF-32 output (worst case: same number of codepoints as bytes)
        let mut utf32_buffer = vec![0u32; haystack.len()];
        measure_throughput(
            "utf8-iterate/simdutf::convert_utf8_to_utf32",
            ReportAs::Bytes,
            budget,
            || {
                let haystack_bytes = black_box(haystack);
                let len = unsafe {
                    simdutf::convert_utf8_to_utf32(
                        haystack_bytes.as_ptr(),
                        haystack_bytes.len(),
                        utf32_buffer.as_mut_ptr(),
                    )
                };
                let mut sum: u32 = 0;
                for value in &utf32_buffer[..len] {
                    sum = sum.wrapping_add(*value);
                }
                black_box(sum);
                WorkUnits::bytes(haystack_length)
            },
        );
    }

    // Benchmark for stdlib UTF-8 character iteration.
    measure_throughput("utf8-iterate/std::chars", ReportAs::Bytes, budget, || {
        // Safety: the tokenization corpora are valid UTF-8 text; `from_utf8_unchecked` skips
        // re-validation on the hot path so the benchmark measures iteration, not UTF-8 checking.
        let haystack_str = black_box(unsafe { std::str::from_utf8_unchecked(haystack) });
        let mut sum: u32 = 0;
        for character in haystack_str.chars() {
            sum = sum.wrapping_add(character as u32);
        }
        black_box(sum);
        WorkUnits::bytes(haystack_length)
    });
}

/// Benchmarks locating the byte offset of the Nth UTF-8 codepoint.
///
/// - `stringzilla::find_nth_utf8()`: SIMD-accelerated scan to the Nth codepoint
/// - `std::str::char_indices().nth()`: scalar decode-and-count baseline
///
/// We target the *last* codepoint, so every implementation scans the whole buffer once —
/// a fair workload whose throughput is simply the input size.
fn bench_find_nth_utf8(budget: &BenchBudget, haystack: &[u8], _needles: &BytesCowsAuto) {
    let haystack_str = match std::str::from_utf8(haystack) {
        Ok(text) => text,
        Err(_) => {
            eprintln!("Warning: Haystack is not valid UTF-8, skipping find-nth-utf8 benchmarks");
            return;
        }
    };

    let codepoint_count = sz::count_utf8(haystack);
    if codepoint_count == 0 {
        return;
    }
    let last_index = codepoint_count - 1;

    let haystack_length = haystack.len() as u64;

    // Benchmark for StringZilla's SIMD `find_nth_utf8`.
    measure_throughput(
        "find-nth-utf8/stringzilla::find_nth_utf8",
        ReportAs::Bytes,
        budget,
        || {
            let offset = sz::find_nth_utf8(black_box(haystack), last_index);
            black_box(offset);
            WorkUnits::bytes(haystack_length)
        },
    );

    // Benchmark for the stdlib scalar baseline.
    measure_throughput(
        "find-nth-utf8/std::char_indices.nth",
        ReportAs::Bytes,
        budget,
        || {
            let offset = black_box(haystack_str)
                .char_indices()
                .nth(last_index)
                .map(|(byte_offset, _)| byte_offset);
            black_box(offset);
            WorkUnits::bytes(haystack_length)
        },
    );
}
fn main() {
    install_panic_hook();
    log_stringzilla_metadata();

    // Load the dataset defined by the environment variables
    let tape = load_dataset().unwrap_nice();

    // Get the parent data directly from the tape (zero-copy)
    let haystack = tape.parent();
    let needles = &tape;

    let budget = BenchBudget::from_env(3.0, 20.0);

    println!("# tokenize-whitespace");
    bench_tokenize_whitespace(&budget, haystack, needles);

    println!("# tokenize-newlines");
    bench_tokenize_newlines(&budget, haystack, needles);

    println!("# tokenize-words-tr29");
    bench_tokenize_words_tr29(&budget, haystack, needles);

    println!("# utf8-length");
    bench_utf8_length(&budget, haystack, needles);

    println!("# utf8-iterate");
    bench_utf8_iterate(&budget, haystack, needles);

    println!("# find-nth-utf8");
    bench_find_nth_utf8(&budget, haystack, needles);
}
